logger/analytics/analyticslogger.go

```diff
@@ -89,22 +127,69 @@ func (al *Logger) Write(bs []byte) (int, error) {
 		return 0, err
 	}
 	bs = append(bs, '\n')
-	r := retrier.New(retrier.ExponentialBackoff(5, 100*time.Millisecond), RequestErrorClassifier{})
-	if err := r.Run(func() error {
-		_, err := al.fhAPI.PutRecord(&firehose.PutRecordInput{
-			DeliveryStreamName: aws.String(al.fhStream),
-			Record:             &firehose.Record{Data: bs},
-		})
-		if err != nil {
-			return err
-		}
-		return nil
-	}); err != nil {
-		return 0, fmt.Errorf("PutRecords: %v", err)
+	al.mu.Lock()
+	al.batchBytes += len(bs)
+	al.batch = append(al.batch, &firehose.Record{Data: bs})
+	shouldSendBatch := len(al.batch) == al.maxBatchRecords || al.batchBytes > int(0.9*float64(al.maxBatchBytes))
+	if shouldSendBatch {
+		batch := al.batch
+		al.batch = []*firehose.Record{}
+		al.batchBytes = 0
+		// be careful not to send al.batch, since we will unlock before we finish sending the batch
+		go sendBatch(batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
 	}
+	al.mu.Unlock()
 	return len(bs), nil
 }
 
+// Close flushes all logs to Firehose.
+func (al *Logger) Close() error {
+	al.mu.Lock()
+	if len(al.batch) > 0 {
+		batch := al.batch
+		al.batch = []*firehose.Record{}
+		al.batchBytes = 0
+		// be careful not to send al.batch, since we will unlock before we finish sending the batch
+		go sendBatch(batch, al.fhAPI, al.fhStream, time.Now().Add(timeoutForSendingBatches))
```

Comment:
IMO this should be a synchronous write. `Closer` is a pretty vague interface but I'd expect everything to be written after `Close` returns.

One example edge case:

```golang
func main() {
	l := logger.New()

	// write a bunch of stuff
	l.Write(...)

	if err := l.Close(); err != nil {
		log.Fatal(err)
	}

	// program can exit before all writes actually finish
}
```

Reply:
ah yeah i forgot to highlight this. Definitely need a waitgroup to count outstanding `sendbatch`s, and something like a timeout to guarantee that `sendBatch` completes. Then you can `Wait` on the waitgroup in `Close`